sympy/printing/tests/test_smtlib.py

```diff
@@ -0,0 +1,430 @@
+        my_fun1(x),
+        symbol_table={my_fun1: Callable[[bool], bool]},
+    ) == "(declare-const x Bool)\n" \
+         "(declare-fun f1 (Bool) Bool)\n" \
+         "(assert (f1 x))"
+
+    assert smtlib_code(
+        Eq(my_fun1(x, z), y),
+        symbol_table={my_fun1: Callable[[int, bool], bool]},
+    ) == "(declare-const x Int)\n" \
+         "(declare-const y Bool)\n" \
+         "(declare-const z Bool)\n" \
+         "(declare-fun f1 (Int Bool) Bool)\n" \
+         "(assert (= (f1 x z) y))"
+
+    assert smtlib_code(
+        Eq(my_fun1(x, z), y),
+        symbol_table={my_fun1: Callable[[int, bool], bool]},
+        known_functions={my_fun1: "MY_KNOWN_FUN", Eq: '=='}
+    ) == "(declare-const x Int)\n" \
+         "(declare-const y Bool)\n" \
+         "(declare-const z Bool)\n" \
+         "(assert (== (MY_KNOWN_FUN x z) y))"
+
+    assert smtlib_code(
+        Eq(my_fun1(x, z), y),
+        known_functions={my_fun1: "MY_KNOWN_FUN", Eq: '=='}
+    ) == "(declare-const x Real)\n" \
+         "(declare-const y Real)\n" \
+         "(declare-const z Real)\n" \
+         "(assert (== (MY_KNOWN_FUN x z) y))"
+
+
+def test_Pow():
+    assert smtlib_code(x ** 3, auto_declare=False) == "(pow x 3)"
+    assert smtlib_code(x ** (y ** 3), auto_declare=False) == "(pow x (pow y 3))"
+    assert smtlib_code(x ** Rational(2, 3), auto_declare=False) == '(pow x (/ 2 3))'
+
+    a = Symbol('a', integer=True)
+    b = Symbol('b', real=True)
+    c = Symbol('c')
+
+    def g(x): return 2 * x
+
+    # if x=1, y=2, then expr=2.333...
+    expr = 1 / (g(a) * 3.5) ** (a - b ** a) / (a ** 2 + b)
+
+    assert smtlib_code([
+        Eq(a < 2, c),
+        Eq(b > a, c),
+        c & True,
+        Eq(expr, 2 + Rational(1, 3))
+    ]) == '(declare-const a Int)\n' \
+          '(declare-const b Real)\n' \
+          '(declare-const c Bool)\n' \
+          '(assert (= (< a 2) c))\n' \
+          '(assert (= (> b a) c))\n' \
+          '(assert c)\n' \
+          '(assert (= ' \
+          '(* (pow (* 7. a) (+ (pow b a) (* -1 a))) (pow (+ b (pow a 2)) -1)) ' \
+          '(/ 7 3)' \
+          '))'
+
+    assert smtlib_code(
+        Mul(-2, c, Pow(Mul(b, b, evaluate=False), -1, evaluate=False), evaluate=False)
+    ) == '(declare-const b Real)\n' \
+         '(declare-const c Real)\n' \
+         '(* -2 c (pow (* b b) -1))'
+
+
+def test_basic_ops():
+    assert smtlib_code(x * y, auto_declare=False) == "(* x y)"
+    assert smtlib_code(x + y, auto_declare=False) == "(+ x y)"
+    # todo: implement re-write, currently does '(+ x (* -1 y))' instead
+    # assert smtlib_code(x - y, auto_declare=False) == "(- x y)"
+    assert smtlib_code(-x, auto_declare=False) == "(* -1 x)"
+
+
+def test_quantifier_extensions():
+    from sympy.logic.boolalg import Boolean
+    from sympy.concrete.expr_with_limits import ExprWithLimits
+    from sympy import Interval, Tuple
+
+    # start For-all quantifier class example
+    class ForAll(Boolean, ExprWithLimits):
```

Comment:
Would it not make sense for this to be a public SymPy class? Or is that just something for a future PR.

If so why does this subclass `ExprWithLimits` which is the superclass for `Sum` and `Integral`?

In general I think that simultaneously subclassing Boolean and Expr is a mistake (the fact that Symbol already does this is a significant design failure).

Reply:
> Would it not make sense for this to be a public SymPy class? Or is that just something for a future PR.

I want to leave implementing Quantifiers outside the scope of this PR because there are multiple attempts already working on it (https://github.com/sympy/sympy/pull/17069 and https://github.com/sympy/sympy/pull/7608). This example class is purely a wrapper to demonstrate how the SMT Printer may be extended. It does not contain any symbolic logic that would make it useful to include in sympy.

> If so why does this subclass ExprWithLimits which is the superclass for Sum and Integral?

Like sum and integral, quantifiers also create bound variables with limits and operate over expressions.
E.g. `∫ f(a)` from `a=0` to `a=100` given `f(x)=...` can evaluate to some number or DNC.
E.g. `∀a∈[0,100], f(-a)=-f(a)` given `f(x)=...` can be proven true, false, or unknown.

Both these expressions have a "bound" variable `a`, place limits on `a` (0 to 100), and also require an expression to evaluate for each `a` within the bounds. Thus, I tag them both as `ExprWithLimits`.
Since `∀a∈...` can be proven true or false, I also tag it as `Boolean`.

>In general I think that simultaneously subclassing Boolean and Expr is a mistake (the fact that Symbol already does this is a significant design failure).

Sounds good, but I'm not sure a good way to fix this. Is there a path forward you suggest? If other classes in Sympy already do this then its a refactoring issue which should be outside the scope of this PR.

Reply:
I would consider those PRs to be stale. Don't let their existence stop you from working on adding quantifiers. 

Adding the ability to specify a symbol and a domain for a quantifier sounds useful, although for the logic module we would also want just general quantifiers untied to sets. However, I don't think ExprWithLimits makes sense. A lot of things on that class don't make sense for quantifiers (like reversing limits), and in general, a quantifier can be made over any set, not just an interval.

Either way, if all you are doing here is testing an extension, then I would make this very simple, like just `class ForAll(Boolean)` with the one `_smt` method. Really all you're testing here is that `_printmethod` works, and that's already pretty well tested for the Printer superclass, so it's not necessary to test that much here (tbh I would even be fine with this test not existing at all).

Reply:
>I would make this very simple, like just class ForAll(Boolean) with the one _smt method.

I understand why `ExprWithLimits` wasn't the best choice now. I'll fix the test as you describe.

However, I disagree that all I'm testing is `_printmethod`. The goal of this test is to verify the initialization and completeness of `_known_types`, `_known_constants`, and `_known_functions`. It exists to make sure anybody who modifies `smtlib_code` (symbol table + type-inference) does not corrupt these data structures. It also tests the printer's auto-declaration code on "local" / bounded variables. E.g. `∀a∈...` should not create a `(declare-const a...`.

> A lot of things on that class don't make sense for quantifiers (like reversing limits), and in general, a quantifier can be made over any set, not just an interval.

True, I was only thinking of the ways I've seen quantifiers used in SMT.